sklearn/metrics/_pairwise_distances_reduction.pyx.tp

```diff
@@ -244,15 +247,593 @@ cdef class PairwiseDistancesReduction:
+              Each thread processes all the chunks of X in turn. This strategy is
+              a sequence of embarrassingly parallel subtasks (the inner loop on Y
+              chunks) with intermediate datastructures synchronisation at each
+              iteration of the sequential outer loop on X chunks.
+
+              - 'auto' relies on a simple heuristic to choose between
+              'parallel_on_X' and 'parallel_on_Y': when `X.shape[0]` is large enough,
+              'parallel_on_X' is usually the most efficient strategy. When `X.shape[0]`
+              is small but `Y.shape[0]` is large, 'parallel_on_Y' brings more opportunity
+              for parallelism and is therefore more efficient despite the synchronization
+              step at each iteration of the outer loop on chunks of `X`.
+
+              - None (default) looks-up in scikit-learn configuration for
+              `pairwise_dist_parallel_strategy`, and use 'auto' if it is not set.
+
+        return_distance : boolean, default=False
+            Return distances between each X vector and its
+            argkmin if set to True.
+
+        Returns
+        -------
+            If return_distance=False:
+              - argkmin_indices : ndarray of shape (n_samples_X, k)
+                Indices of the argkmin for each vector in X.
+
+            If return_distance=True:
+              - argkmin_distances : ndarray of shape (n_samples_X, k)
+                Distances to the argkmin for each vector in X.
+              - argkmin_indices : ndarray of shape (n_samples_X, k)
+                Indices of the argkmin for each vector in X.
+
+        Notes
+        -----
+            This public classmethod is responsible for introspecting the arguments
+            values to dispatch to the proper implementations of
+            :meth:`PairwiseDistancesArgKmin` given the types.
+
+            All temporarily allocated datastructures necessary for the concrete
+            implementation are therefore freed when this classmethod returns.
+
+            This allows decoupling the interface entirely from the
+            implementation details whilst maintaining RAII.
+        """
+{{for distance_suffix, bitness, DTYPE_t, DTYPE, need_upcast in implementation_specific_values}}
+        if X.dtype == Y.dtype == np.float{{bitness}}:
+            return PairwiseDistancesArgKmin{{bitness}}.compute(
+                X=X,
+                Y=Y,
+                k=k,
+                metric=metric,
+                chunk_size=chunk_size,
+                metric_kwargs=metric_kwargs,
+                strategy=strategy,
+                return_distance=return_distance,
+            )
+{{endfor}}
+        raise ValueError(
+            f"Datasets must both be of np.float64 or np.float32 dtype. "
+            f"Currently: X.dtype={X.dtype} and Y.dtype={Y.dtype}."
+        )
+
+cdef class PairwiseDistancesRadiusNeighborhood(PairwiseDistancesReduction):
```

Comment:
Note to reviewers: similarly `PairwiseDistancesRadiusNeighborhood` is just an interface now which dispatch to the correct dtype-specific implementation at runtime.